src/RequestGroupMan.cc

```diff
@@ -474,11 +484,43 @@ createInitialCommand(const std::shared_ptr<RequestGroup>& requestGroup,
 void RequestGroupMan::fillRequestGroupFromReserver(DownloadEngine* e)
 {
   removeStoppedGroup(e);
-  if (static_cast<size_t>(maxSimultaneousDownloads_) <= numActive_) {
+
+  int maxSimultaneousDownloads=maxSimultaneousDownloads_;
+
+  if(optimizeSimultaneousDownloads_) {
+    // optimize method is requested, so gauge current speed
+    int currentSpeed=getNetStat().calculateDownloadSpeed();
+    if(currentSpeed>0) {
+      // apply the rule
+      int speed=currentSpeed;
+      if((maxOverallDownloadSpeedLimit_>0) && (speed>maxOverallDownloadSpeedLimit_)) {
+        speed=maxOverallDownloadSpeedLimit_;
+      }
+      maxSimultaneousDownloads=ceil(
+        optimizeSimultaneousDownloadsCoeffA_
+        +optimizeSimultaneousDownloadsCoeffB_ * log10(speed*8./1000000.)
+      );
+      if((maxSimultaneousDownloads_>0)&&(maxSimultaneousDownloads>maxSimultaneousDownloads_)) {
+        maxSimultaneousDownloads=maxSimultaneousDownloads_;
+      }
+      if(maxSimultaneousDownloads<1) {
```

Comment:
We can first check this condition before the above if statement to make the intention clear.  Something like:
```cpp
if (maxSimultaneousDownloads < 1) {
  maxSimultaneousDownloads = 1;
} else if (maxSimultaneousDownloads > maxSimultaneousDownloads_) {
  maxSimultaneousDownlaods = maxSimultaneousDownloads_;
}
```
or just go in one-liner:
```cpp
maxSimultaneousDownloads = std::min(std::max(1, maxSimultaneousDownloads), maxSimultaneousDownloads_);
```